Clear last_used_directory when download.location.directory changes

mehalter opened the pull request:
If there is a change in the configuration download location, this new location is used instead of the last used directory. Fixes issue #4674

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/qutebrowser/qutebrowser/4675)
<!-- Reviewable:end -->

toofar commented:
I think it would complicate the code less to have a change filter to unset `last_used_directory` when `downloads.location.directory` changes instead. Like how the proxy thing [does it](https://github.com/qutebrowser/qutebrowser/blob/3e816f5/qutebrowser/browser/network/proxy.py#L38). And since I don't think we should be calling config stuff at import time you would also have to add a `init()` method and call that from `app.py:init_modules()`.
You would also need to add a test. It looks like current tests around that behaviour are [here](https://github.com/qutebrowser/qutebrowser/blob/3e816f5/tests/end2end/features/downloads.feature#L513). I think everything you need to write one for this is there already.

mehalter commented:
Implemented this fix for issue #4674 as described in the previous comment to make it more maintainable and consistent with the implementation of other features in the code base. Also added a test for the new change.

jgkamat commented on qutebrowser/browser/downloads.py:
```diff
@@ -46,11 +46,6 @@ class ModelRole(enum.IntEnum):
 
-
-# Remember the last used directory
-last_used_directory = None
```
I'm not sure if this is standard practice, but I would prefer to leave this initialization here, if only to show that such a variable exists in this file (instead of inferring it from the globals). 

It can also cause weird issues if, for example, `init` isn't run before other methods.

mehalter replied:
I added this initialization back in

jgkamat commented on tests/end2end/features/downloads.feature:
```diff
@@ -522,6 +522,17 @@ Feature: Downloading things from a website.
+        And I open data/downloads/download.bin without waiting
+        And I wait for the download prompt for "*/download.bin"
+        And I run :prompt-accept (tmpdir)(dirsep)downloads(dirsep)subdir
+        And I set downloads.location.directory to (tmpdir)
```
`(tmpdir)` isn't expanded here:
```
Traceback (most recent call last):
  File "/home/jay/Code/qutebrowser/qutebrowser/config/configcommands.py", line 55, in _handle_config_error
    yield
  File "/home/jay/Code/qutebrowser/qutebrowser/config/configcommands.py", line 135, in set
    save_yaml=not temp)
  File "/home/jay/Code/qutebrowser/qutebrowser/config/config.py", line 461, in set_str
    converted = opt.typ.from_str(value)
  File "/home/jay/Code/qutebrowser/qutebrowser/config/configtypes.py", line 251, in from_str
    self.to_py(value)  # for validation
  File "/home/jay/Code/qutebrowser/qutebrowser/config/configtypes.py", line 1452, in to_py
    value, "must be a valid directory!")
qutebrowser.config.configexc.ValidationError: Invalid value '(tmpdir)' - must be a valid directory!

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jay/Code/qutebrowser/qutebrowser/commands/runners.py", line 329, in run_safely
    self.run(text, count)
  File "/home/jay/Code/qutebrowser/qutebrowser/commands/runners.py", line 308, in run
    result.cmd.run(self._win_id, args, count=count)
  File "/home/jay/Code/qutebrowser/qutebrowser/commands/command.py", line 551, in run
    self.handler(*posargs, **kwargs)
  File "/home/jay/Code/qutebrowser/qutebrowser/config/configcommands.py", line 135, in set
    save_yaml=not temp)
  File "/usr/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/jay/Code/qutebrowser/qutebrowser/config/configcommands.py", line 57, in _handle_config_error
    raise cmdutils.CommandError(str(e))
qutebrowser.api.cmdutils.CommandError: Invalid value '(tmpdir)' - must be a valid directory!
19:06:20.657 ERROR    message    message:error:58 Invalid value '(tmpdir)' - must be a valid directory!
```

You should be able to fix it by creating a one-off command (although I'm not sure if that's the best way), or by making this a unit test. See `test_downloads_bdd.py` for the command definitions here (searching for `(tmpdir)` in that file is helpful)

mehalter replied:
I made this a one-off set command instead to fix the tmpdir expansion

mehalter commented:
Fixed the linting issues and it looks like the tests all pass :100:

mehalter commented:
No problem, I haven't contributed to this project yet and am getting adjusted to the development workflow for the project. Thanks for the guidance.

The-Compiler commented:
Meh, Travis is kinda broken and AppVeyor looks okay, let's assume it's fine :laughing:
